Добавлены SplashScreen и HomeScreen.

shvetsovalex opened the pull request.

shvetsovalex committed:
Добавлены SplashScreen и HomeScreen.

lakutkin commented on SplashScreen/app/src/main/AndroidManifest.xml:
```diff
@@ -1,10 +1,27 @@
+        android:supportsRtl="true"
+        android:theme="@style/AppTheme" >
+        <activity
+            android:name=".MainActivity"
```
Если это сплэш скрин, то неудачное название класса активити.

shvetsovalex replied:
Название класса должно было быть SplashScreen

lakutkin commented on SplashScreen/app/src/main/AndroidManifest.xml:
```diff
@@ -1,10 +1,27 @@
+        <activity
+            android:name=".HomeActivity"
+            android:label="@string/title_activity_home"
+            android:theme="@style/AppTheme.NoActionBar" >
```
Обе активити имеют стиль NoActionBar. В таких случаях имеет смысл объявить эту тему на все приложение (в application тэге) и не повторяться.

shvetsovalex replied:
android:theme="@style/NoActionBar" надо было объявить в тэге application

lakutkin commented on SplashScreen/app/src/main/java/com/example/splashscreen/HomeActivity.java:
```diff
@@ -0,0 +1,30 @@
+        ArrayAdapter<String> adapter = new ArrayAdapter<String>(this,
+                android.R.layout.simple_list_item_1, names);
+
+        // присваиваем адаптер списку
```
В чем смысл комментировать каждую строчку кода?

shvetsovalex replied:
Когда писал,чтобы лучше запомнить, забыл удалить,  впредь не буду

lakutkin commented on SplashScreen/app/src/main/java/com/example/splashscreen/MainActivity.java:
```diff
@@ -0,0 +1,34 @@
+
+public class MainActivity extends Activity {
+    // Splash screen timer
+    private static int SPLASH_TIME_OUT = 2000;
```
Комментарий к константе неправильный. Это не таймер, это таймаут. И поять же, комментарий ничего не добавляет нового, из названия константы и так понятно, что это такое (и так и должно быть)

shvetsovalex replied:
Да, комментарий лишний опять же

lakutkin commented on SplashScreen/app/src/main/java/com/example/splashscreen/MainActivity.java:
```diff
@@ -0,0 +1,34 @@
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.splash);
+        if(splash) {
```
а этот if работает?

shvetsovalex replied:
If работает по крайней мере на эмуляторе Genymotion. На реальном android не тестировал, на эмуляторе же всё работает хорошо.

lakutkin replied:
Вопрос не в этом. Есть ли хоть один сценарий, где if (splash) будет равно false?

shvetsovalex replied:
Будет, когда произойдет поворот экрана. Активити завершает свой жизненный цикл и вызывается метод onDestroy().

lakutkin commented on SplashScreen/app/src/main/res/layout/activity_home.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="fill_parent"
```
fill_parent - deprecated. Надо использовать match parent везде.

shvetsovalex replied:
Да, прочитал, fill parent устаревшее, но его совместимость пока оставили для поддержки

lakutkin commented on SplashScreen/app/src/main/res/layout/activity_home.xml:
```diff
@@ -0,0 +1,18 @@
+        android:layout_width="fill_parent"
+        android:layout_height="wrap_content"
+        android:text="@string/home_screen"
+        android:gravity="center_horizontal">
```
В чем отлииче android:gravity от android:layoutGravity? Почему в данном случае принято решение использовать именно gravity?

shvetsovalex replied:
Отличие в том, что используя gravity, идет выравнивание текста внутри TextView, а при использовании layoutGravity выравнивание всего тэга относительно родителя. Здесь уместнее наверное было бы использовать layout_gravity и layout_width="wrap_content"

lakutkin replied:
:+1:

lakutkin commented on SplashScreen/app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,16 @@
+    <string name="item_7">item 7</string>
+    <string name="item_8">item 8</string>
+    <string name="item_9">item 9</string>
+
```
зачем нужны эти отдельные строки?

shvetsovalex replied:
Не знал, как правильно реализовать ресурсы. Решил,что все строковые константы должны лежать в strings.xml, но видимо это не так.

lakutkin replied:
А, понял о чем ты. Из-за одинаковых имен переменных и их значений, я не заметил, что в arrays.xml там ссылки на strings.xml. Это ок.

lakutkin commented:
Самый главный вопрос - тестировал ли ты сплеш скрин? Багов не обнаружил?

shvetsovalex commented:
Нет, не обнаружил багов. Тестировал на Genymotion, на образе андроида версии 4.2.2. Через 2 секунды исчезает сплеш скрин и появляется домашний экран со списком. При поворотах сплеш скрина новые интенты не создаются.

lakutkin commented on SplashScreen/app/src/main/java/com/example/splashscreen/MainActivity.java:
```diff
@@ -0,0 +1,34 @@
+public class MainActivity extends Activity {
+    // Splash screen timer
+    private static int SPLASH_TIME_OUT = 2000;
+    private static boolean splash = true;
```
1) плохое название переменной
2) в данном конкретном случае статическая переменная сойдет. В общем случае - нет. Нужно убрать static и найти другой способ. 

Как найдешь другой (правильный) способ, обсудим, почему текущий вариант в общем случае неправильный.

lakutkin replied:
Хоть в данном случае статическая переменная и сойдет, реализация неправильная. 
Тесты
1) запустить приложение, дождаться экрана со списком. Нажать back. Достать приложение из бэкграунда. 
2) запустить приложение, пока отображается splash screen, увести в бэкграунд кнопкой home.

shvetsovalex committed:
gravity заменен на layout_gravity, использовано более подходящее имя класса(SplashScreen).
Также для получения информации о том, создавалась ли активити используется массив типа Bandle

shvetsovalex commented:
Исправил ошибки, тесты при предыдущем способе не прошли. Сделал через массив Bundle

lakutkin commented:
А при текущем способе тесты проходят?

shvetsovalex commented:
При текущем способе, при нажатии кнопки Home и дальнейшем разворачивании появляется Home Screen, а при нажатии кнопки назад, появляется снова splashscreen, так как активити HomeScreen уничтожается как самая верхняя из стека таска. Так и должно быть?

lakutkin commented:
Просьба описывать ситуацию предельно подробно, т.к. из такого описания непонятно, в чем твой вопрос.

> 1)  при нажатии кнопки Home и дальнейшем разворачивании появляется Home Screen

При нажатии Home с какого экрана? Что значит при дальнейшем разворачивании? В чем был мой изначальный вопрос?

> 2) при нажатии кнопки назад, появляется снова splashscreen, так как активити HomeScreen уничтожается как самая верхняя из стека таска. 

Это поведение верное.

shvetsovalex commented:
При нажатии Home из SplashSсreen приложение уходит из фокуса а потом, при развороте, появляется HomeScreen, так как уже проходит 2 секунды после первого появления SplashScreen. Если поставить таймаут 5 секунд и свернуть и развернуть приложение за 3 секунды, то развернется снова SplashScreen, повисит 2 секунды и появится HomeScreen

lakutkin commented:
У меня на симуляторе поведение другое, такое, какое оно и должно быть, глядя на код.

После нажатия на Home из SplashScreen, приложение уходит в бэкграунд, ждем 5 секунд, и открывается HomeScreen.

shvetsovalex commented:
После нажатия на Home из SplashScreen, приложение уходит в бэкграунд, ждем 5 секунд, и открывается HomeScreen. Сам HomeScreen открываться не должен, я понял,но когда разворачиваем, должен открываться HomeScreen или снова SplashScreen?

shvetsovalex commented:
Есть такая мысль, в коде метода run объекта runnable передать активити HomeScreen ссылку на массив Bandle. Если за время отображения SplashScreen активити SplashScreen будет принудительно уничтожена, то вызовется метод onSaveInstanceState, в котором в массив Bandle, ссылку на который передаю в активити HomeScreen, устанавливаю флаг, что приложение свернуто(переменная boolean). Соответственно, когда после таймаута идет вызов HomeScreen через Интент в него же передаётся ссылка на Bandle, в котором есть информация о том, свернуто приложение или нет. Это получилось сделать. Но проблема в том, что я не знаю как программно свернуть HomeActivity сразу после создания. Это возможно?

shvetsovalex committed:
Сделал через removeCallback.

shvetsovalex committed:
При сворачивании приложения кнопками Back и Home на SplashScreen HomeScreen сам не появляется. При многократных поворотах экрана SplasScreen показывается только 2 секунды.

lakutkin commented:
:+1:
